packages/contracts-bridge/contracts/BridgeRouter.sol

```diff
@@ -136,49 +139,102 @@ contract BridgeRouter is Version0, Router {
-            // query token contract for details and calculate detailsHash
-            _detailsHash = BridgeMessage.getDetailsHash(
-                _t.name(),
-                _t.symbol(),
-                _t.decimals()
-            );
-        } else {
-            // if the token originates on a remote chain,
-            // burn the representation tokens on this chain
-            _t.burn(msg.sender, _amount);
-            _detailsHash = _t.detailsHash();
-        }
-        // format Transfer Tokens action
+        // debit tokens from the sender
+        (bytes29 _tokenId, bytes32 _detailsHash) = _debitTokens(
+            _token,
+            _amount
+        );
+        // format Transfer message
         bytes29 _action = BridgeMessage.formatTransfer(
             _recipient,
             _amount,
             _detailsHash
         );
-        // get the tokenID
-        (uint32 _domain, bytes32 _id) = tokenRegistry.getTokenId(_token);
-        bytes29 _tokenId = BridgeMessage.formatTokenId(_domain, _id);
-        // send message to remote chain via Nomad
-        Home(xAppConnectionManager.home()).dispatch(
-            _destination,
-            _remote,
-            BridgeMessage.formatMessage(_tokenId, _action)
-        );
+        // send message to destination chain bridge router
+        _sendTransferMessage(_destination, _tokenId, _action);
         // emit Send event to record token sender
         emit Send(_token, msg.sender, _destination, _recipient, _amount, false);
     }
 
+    /**
+     * @notice Send tokens to a hook on the remote chain
+     * @param _token The token address
+     * @param _amount The token amount
+     * @param _destination The destination domain
+     * @param _hook The hook contract on the remote chain
+     * @param _extraData Extra data that will be passed to the hook for
+     *        execution
+     */
+    function _xsend(
+        address _token,
+        uint256 _amount,
+        uint32 _destination,
+        bytes32 _hook,
+        bytes calldata _extraData
+    ) internal {
+        // debit tokens from the Connext
+        (bytes29 _tokenId, bytes32 _detailsHash) = _debitTokens(
+            _token,
+            _amount
+        );
+        // format Connext transfer message
+        bytes29 _action = BridgeMessage.formatTransferToHook(
+            _hook,
+            _amount,
+            _detailsHash,
+            _extraData
+        );
+        // send message to destination chain bridge router
+        _sendTransferMessage(_destination, _tokenId, _action);
+        // TODO: emit special event?
+    }
+
+    /**
+     * @notice Send tokens to a hook on the remote chain
+     * @param _token The token address
+     * @param _amount The token amount
+     * @param _destination The destination domain
+     * @param _hook The hook contract on the remote chain
+     * @param _extraData Extra data that will be passed to the hook for
+     *        execution
+     */
+    function xsend(
```

Comment:
typechain will be ugly if we have 2 functions with the same name. we should probably only have the `bytes32` version